javascript_nodejs/50.contoso-cafe-bot/dialogs/main/index.js

```diff
@@ -0,0 +1,260 @@
+            this.addDialog(new WhatCanYouDoDialog());
+            this.addDialog(new CancelDialog());
+            this.addDialog(new FindCafeLocationsDialog());
+            this.addDialog(new QnADialog(botConfig, this.userProfilePropertyAccessor));
+            this.addDialog(new WhoAreYouDialog(botConfig, this.userProfilePropertyAccessor, conversationState));
+        }
+        /**
+         * Override onDialogBegin 
+         * 
+         * @param {Object} dc dialog context
+         * @param {Object} options dialog turn options
+         */
+        async onDialogBegin(dc, options) {
+            // Override default begin() logic with bot orchestration logic
+            return await this.mainDispatch(dc);
+        }
+        /**
+         * Override onDialogContinue
+         * 
+         * @param {Object} dc dialog context
+         */
+        async onDialogContinue(dc) {
+            // Override default continue() logic with bot orchestration logic
+            return await this.mainDispatch(dc);
+        }
+        /**
+         * Main Dispatch 
+         * 
+         * This method examines the incoming turn property to determine  
+         * 1. If the requested operation is permissible - e.g. if user is in middle of a dialog, 
+         *     then an out of order reply should not be allowed.
+         * 2. Calls any oustanding dialogs to continue
+         * 3. If results is no-match from outstanding dialog .OR. if there are no outstanding dialogs,
+         *    decide which child dialog should begin and start it
+         * 
+         * @param {Object} dc dialog context
+         */
+        async mainDispatch(dc) {
+            // get on turn property through the property accessor
+            const onTurnProperty = await this.onTurnPropertyAccessor.get(dc.context);
+            
+            // Evaluate if the requested operation is possible/ allowed.
+            const reqOpStatus = await this.isRequestedOperationPossible(dc, onTurnProperty.intent);
+            if (!reqOpStatus.allowed) {
+                await dc.context.sendActivity(reqOpStatus.reason);
+                // Nothing to do here. End main dialog.
+                return await dc.end(); 
+            }
+            
+            let dialogTurnResult;
+
+            // continue outstanding dialogs
+            dialogTurnResult = await dc.continue();
+
+            // This will only be empty if there is no active dialog in the stack.
+            if (!dc.context.responded) {
+                // No one has responded so start the right child dialog.
+                dialogTurnResult = await this.beginChildDialog(dc, onTurnProperty);
+            }
+
+            if(dialogTurnResult === undefined) return await dc.end();
+
+            // Examine result from dc.continue() or from the call to beginChildDialog().
+            switch (dialogTurnResult.status) {
+                case DialogTurnStatus.complete: {
```

Comment:
This case should be broken out in to a separate function to improve readability.